### PR TITLE
mds: PurgeQueue.cc fix for 32bit compilation

### PR DESCRIPTION
**Pull request by a16bitsysop**
files_high_water is defined as uint64_t but when compiling on 32bit these max functions
fail as they are both not considered uint64_t by gcc 10 even though they are

  files_high_water = std::max(files_high_water,
                               static_cast<uint64_t>(in_flight.size()));

Fixes: https://tracker.ceph.com/issues/50707

Signed-off-by: Duncan Bellamy <dunk@denkimushi.com>

## Checklist
- [X] References tracker ticket
- [X] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/mds/PurgeQueue.cc [hunk 1]
```diff
@@ -594,7 +594,7 @@ void PurgeQueue::_execute_item(
 
   in_flight[expire_to] = item;
   logger->set(l_pq_executing, in_flight.size());
-  files_high_water = std::max(files_high_water,
+  files_high_water = std::max(static_cast<uint64_t>(files_high_water),
```
Comment (tchaikov) [inline src/mds/PurgeQueue.cc:597]: why? `files_high_water` is already an `uint64_t`.

File: src/mds/PurgeQueue.cc [hunk 2]
```diff
@@ -662,7 +662,7 @@ void PurgeQueue::_execute_item(
     logger->set(l_pq_executing_ops_high_water, ops_high_water);
     in_flight.erase(expire_to);
     logger->set(l_pq_executing, in_flight.size());
-    files_high_water = std::max(files_high_water,
+    files_high_water = std::max(static_cast<uint64_t>(files_high_water),
```
Comment (tchaikov) [inline src/mds/PurgeQueue.cc:665]: ditto.

File: src/mds/PurgeQueue.cc [hunk 3]
```diff
@@ -716,7 +716,7 @@ void PurgeQueue::_execute_item_complete(
 
   in_flight.erase(iter);
   logger->set(l_pq_executing, in_flight.size());
-  files_high_water = std::max(files_high_water,
+  files_high_water = std::max(static_cast<uint64_t>(files_high_water),
```
Comment (tchaikov) [inline src/mds/PurgeQueue.cc:719]: ditto.

File: src/mds/PurgeQueue.cc (hunk 1 above)
Comment (tchaikov) [reply src/mds/PurgeQueue.cc:597]: i just read the description in the tracker. but i still don't understand why GCC fails to figure out the first parameter is actually an `uint64_t`. probably a less confusing change is to have something like
```c++
files_high_water = std::max<uint64_t>(files_high_water,
                                      in_flight.size());
``` 

but still it's weird..

Comment (a16bitsysop) [reply src/mds/PurgeQueue.cc:597]: Log snippet
`no matching function for call to 'max(size_t&, uint64_t)'`

it seems to assigned size_t by gcc?

Comment (a16bitsysop) [reply src/mds/PurgeQueue.cc:597]: Complete failed build log here:
https://gitlab.alpinelinux.org/a16bitsysop/aports/-/jobs/387736/raw

File: src/mds/PurgeQueue.cc (hunk 3 above)
Comment (tchaikov) [reply src/mds/PurgeQueue.cc:719]: could you apply the same change?

Comment (a16bitsysop) [reply src/mds/PurgeQueue.cc]: Done it

**Review by tchaikov — APPROVED** [review]

File: src/mds/PurgeQueue.cc (hunk 1 above)
Comment (tchaikov) [reply src/mds/PurgeQueue.cc]: just read it. assuming you are using v16.2.3, i don't have a clue explaining this. probably it's a GCC bug. but i cannot find any bug report similar to this one
```diff
diff --git a/src/mds/PurgeQueue.h b/src/mds/PurgeQueue.h
index 6e953d3e633..dd678de4088 100644
--- a/src/mds/PurgeQueue.h
+++ b/src/mds/PurgeQueue.h
@@ -245,5 +245,6 @@ private:

   uint64_t ops_high_water = 0;
   uint64_t files_high_water = 0;
+  static_assert(std::is_same_v<decltype(files_high_water), uint64_t>);
 };
 #endif
```
if you are able to reproduce this locally. probably you could give the patch above a try.

Comment (a16bitsysop) [reply src/mds/PurgeQueue.cc]: Yes using 16.2.3.  I can try on a pi4 and see if it compiles

Comment (tchaikov) [reply src/mds/PurgeQueue.cc]: FWIW, #41239 is created to address the warnings in https://gitlab.alpinelinux.org/a16bitsysop/aports/-/jobs/387736/raw.

Comment (a16bitsysop) [reply src/mds/PurgeQueue.cc]: Thanks will add that patch as well

File: src/mds/PurgeQueue.cc [hunk 4]
```diff
@@ -716,19 +716,19 @@ void PurgeQueue::_execute_item_complete(
...
   dout(10) << "in_flight.size() now " << in_flight.size() << dendl;
 
-  uint64_t write_pos = journaler.get_write_pos(); 
-  uint64_t read_pos = journaler.get_read_pos(); 
-  uint64_t expire_pos = journaler.get_expire_pos(); 
-  uint64_t item_num = (write_pos - (in_flight.size() ? expire_pos : read_pos)) 
+  uint64_t write_pos = journaler.get_write_pos();
+  uint64_t read_pos = journaler.get_read_pos();
+  uint64_t expire_pos = journaler.get_expire_pos();
+  uint64_t item_num = (write_pos - (in_flight.size() ? expire_pos : read_pos))
```
Comment (batrick) [inline src/mds/PurgeQueue.cc:727]: This is code churn, please undo these changes.

File: src/mds/PurgeQueue.cc [hunk 5]
```diff
@@ -716,19 +716,19 @@ void PurgeQueue::_execute_item_complete(
...
 		      / purge_item_journal_size;
-  dout(10) << "left purge items in journal: " << item_num 
-    << " (purge_item_journal_size/write_pos/read_pos/expire_pos) now at " 
-    << "(" << purge_item_journal_size << "/" << write_pos << "/" << read_pos 
+  dout(10) << "left purge items in journal: " << item_num
+    << " (purge_item_journal_size/write_pos/read_pos/expire_pos) now at "
+    << "(" << purge_item_journal_size << "/" << write_pos << "/" << read_pos
```
Comment (batrick) [inline src/mds/PurgeQueue.cc:731]: ditto

File: src/mds/PurgeQueue.cc [hunk 6]
```diff
@@ -7,9 +7,9 @@
  *
  * This is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
- * License version 2.1, as published by the Free Software 
+ * License version 2.1, as published by the Free Software
  * Foundation.  See file COPYING.
- * 
+ *
```
Comment (batrick) [inline src/mds/PurgeQueue.cc:12]: ditto

**Review by batrick — CHANGES_REQUESTED** [review]

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-May-11
